torch/distributed/_state_dict_utils.py

```diff
@@ -50,60 +60,97 @@ def _all_gather_sharded_tensor(
     return tensor
 
 
+class CompanionMismatch(Exception):
+    ...
+
+
 def _iterate_state_dict(
     iter_object: Any,
     sharded_tensor_func: Callable,
     dtensor_func: Callable,
+    tensor_func: Callable,
     *,
     pg: Optional[dist.ProcessGroup] = None,
     device: Optional[torch.device] = None,
     cpu_offload: bool = False,
+    companion_obj: Any = None,
     ranks_only: Tuple[int, ...] = tuple(),
+    type_check: bool = True,
 ) -> Dict[str, Any]:
     # TODO: should we use pytree?
     cpu_device = torch.device("cpu")
     if isinstance(iter_object, ShardedTensor):
-        ret = sharded_tensor_func(iter_object, pg, device)
+        ret = sharded_tensor_func(iter_object, pg, device, companion_obj)
     elif isinstance(iter_object, DTensor):
-        ret = dtensor_func(iter_object, pg, device)
+        ret = dtensor_func(iter_object, pg, device, companion_obj)
+    elif isinstance(iter_object, torch.Tensor):
+        ret = tensor_func(iter_object, pg, device, companion_obj)
     elif (
-        isinstance(iter_object, (torch.Tensor, int, float, str)) or iter_object is None
+        isinstance(iter_object, (int, float, str, bytes, io.BytesIO))
+        or iter_object is None
     ):
         ret = iter_object
     elif isinstance(iter_object, dict):
+        if companion_obj is None:
```

Comment:
I think I may be missing something but why create a campanion obj if it is None? Can't we just skip the comparison in that case?

Reply:
This is just a trick to make the comparison easily. Since we will unroll the companion object as well. So instead of check whether it is None and pass None else passing companion[key], I create a dictionary full of None so that I can also write `companion[key]`. But this may be too implicit. I can change this.